File: 295. Find Median from Data Stream.java
class MedianFinder {
    PriorityQueue<Integer> front, back;
    public MedianFinder() {
        front = new PriorityQueue<>((a, b) -> Integer.compare(b, a));
        back = new PriorityQueue<>();
    }
    
    public void addNum(int num) {
        if(front.size() == 0) front.offer(num);
        else if(back.size() == 0) back.offer(num);
        else{
            if(num > back.peek()){
                front.offer(back.poll());
                back.offer(num);
            }
            else front.offer(num);

            if(front.size() - back.size() == 2){
                back.offer(front.poll());
            }
            else if(back.size() - front.size() == 2){
                front.offer(back.poll());
            }
        } 
    }
    
    public double findMedian() {
        if(front.size() == back.size()){
            return ((double)(front.peek() + back.peek()) / 2.0);
        }
        else if(back.size() == 0) return front.peek();
        return Math.min(front.peek(), back.peek());
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */

File: 295. Find Median from Data Stream.cpp
class MedianFinder {
    priority_queue<int> front;
    priority_queue<int, vector<int>, greater<int>> back;
public:
    MedianFinder() {}

    void addNum(int num) {
        if(front.empty()) front.push(num);
        else if(back.empty()) back.push(num);
        else {
            if(num > back.top()) {
                front.push(back.top());
                back.pop();
                back.push(num);
            } else {
                front.push(num);
            }

            if(front.size() - back.size() == 2) {
                back.push(front.top());
                front.pop();
            } else if(back.size() - front.size() == 2) {
                front.push(back.top());
                back.pop();
            }
        }
    }

    double findMedian() {
        if(front.size() == back.size()) {
            return (front.top() + back.top()) / 2.0;
        } else if(back.empty()) {
            return front.top();
        }
        return min(front.top(), back.top());
    }
};